### PR TITLE
Add Flow-level checkpointing with sample-level tracking

### PR DESCRIPTION
**Pull request by shivchander**
## Summary

This PR introduces comprehensive checkpointing functionality for SDG Hub flows, enabling graceful recovery from interruptions and efficient resumption of long-running data generation processes.

### Key Features

- **Flow-level checkpointing**: Automatic saving and loading of generation progress
- **Sample-level tracking**: Fine-grained tracking of completed vs remaining samples
- **Configurable save frequency**: Control how often checkpoints are created during generation
- **Seamless resumption**: Automatically resume from the last checkpoint when restarting
- **Non-intrusive design**: Checkpointing is optional and doesn't affect existing workflows

### Core Components

1. **FlowCheckpointer class** (`src/sdg_hub/core/flow/checkpointer.py`)
   - Manages checkpoint creation, loading, and sample tracking
   - Uses JSONL format for efficient incremental saves
   - Includes metadata tracking for flow validation

2. **Enhanced Flow.generate() method** (`src/sdg_hub/core/flow/base.py`)
   - Added `checkpoint_dir` and `save_freq` parameters
   - Integrates checkpointing seamlessly into existing flow execution
   - Supports both chunked processing and full dataset processing

3. **Comprehensive test coverage** (`tests/flow/test_checkpointer.py`, `tests/flow/test_base.py`)
   - Tests for all checkpointing scenarios
   - Integration tests with multiple blocks
   - Edge case handling (empty datasets, resume scenarios)

### Usage Example

```python
# Enable checkpointing with automatic saves every 100 samples
result = flow.generate(
    dataset, 
    checkpoint_dir="./checkpoints",
    save_freq=100
)

# On restart, automatically resumes from last checkpoint
result = flow.generate(dataset, checkpoint_dir="./checkpoints")
```

### Benefits

- **Fault tolerance**: Recover from system failures or interruptions
- **Cost efficiency**: Avoid re-processing expensive LLM generations
- **Progress monitoring**: Track completion status during long runs
- **Flexible configuration**: Customize checkpointing behavior per use case

## Test Plan

- [x] Unit tests for FlowCheckpointer class
- [x] Integration tests with Flow.generate() method
- [x] Resume functionality testing
- [x] Edge case handling (empty datasets, all samples completed)
- [x] Multi-block flow checkpointing
- [x] Runtime parameters compatibility
- [x] Backwards compatibility (no checkpointing)

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

- **New Features**
  - Flow-level checkpointing for resumable execution with per-sample/chunk saves, progress merging, final flush, and cleanup utilities.
  - Optional chunked processing via configurable save frequency.
  - New validated dataset concatenation helper emitting clearer errors on failure.

- **Refactor**
  - Block-execution logic extracted to improve clarity, error handling, and resume flow.

- **Tests**
  - Comprehensive tests for checkpointing, resumption, chunking, metadata, edge cases, and checkpointer behavior.

- **Documentation**
  - generate() docstring updated to document checkpoint parameters.

### CODE REVIEW

**Review by eshwarprasadS — COMMENTED** [review]
## Critical Issues Review

Found a few areas that could benefit from additional consideration:

**Input Validation (Line 45-46 in checkpointer.py):**
In the  method, could we consider adding validation for  to ensure it's > 0? This might prevent runtime issues when users accidentally pass negative values.

**Duplicate Handling (Lines 267-270 in checkpointer.py):**
In , the set-based approach for finding remaining samples might not preserve duplicate rows in the input dataset. Could we consider using  with an indicator to handle cases where the input dataset contains duplicate entries?

**Logging Accuracy (Lines 146-148 in checkpointer.py):**
In , the log message shows 0 samples because  is cleared in  before logging. Could we capture the count before clearing to make the log more accurate?

**Review by eshwarprasadS — COMMENTED** [review]
Found a few areas that could benefit from additional consideration:

**Input Validation (Line 45-46 in checkpointer.py):**
In the init method, could we consider adding validation for save_freq to ensure it's > 0? This might prevent runtime issues when users accidentally pass negative values.

**Duplicate Handling (Lines 267-270 in checkpointer.py):**
In find_remaining_samples, the set-based approach for finding remaining samples might not preserve duplicate rows in the input dataset. Could we consider using pd.merge() with an indicator to handle cases where the input dataset contains duplicate entries?

**Logging Accuracy (Lines 146-148 in checkpointer.py):**
In save_final_checkpoint, the log message shows 0 samples because pending_samples is cleared in save_checkpoint() before logging. Could we capture the count before clearing to make the log more accurate?

File: src/sdg_hub/core/flow/base.py [hunk 1]
```diff
@@ -380,18 +388,113 @@ def generate(
...
 
-        current_dataset = dataset
         # Merge migrated runtime params with provided ones (provided ones take precedence)
         merged_runtime_params = self._migrated_runtime_params.copy()
         if runtime_params:
             merged_runtime_params.update(runtime_params)
         runtime_params = merged_runtime_params
 
+        # Process dataset in chunks if checkpointing with save_freq
+        if checkpointer and save_freq:
```
Comment (mihirathale98) [inline src/sdg_hub/core/flow/base.py:457]: Refactor suggestion - This conditional processing could be made cleaner by exporting to a function

File: src/sdg_hub/core/flow/checkpointer.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+                    )
+            except Exception as exc:
+                logger.warning(f"Failed to load checkpoint {file_path}: {exc}")
+
+        if not datasets:
+            return None
+
+        return concatenate_datasets(datasets)
+
+    def _find_remaining_samples(
```
Comment (mihirathale98) [inline src/sdg_hub/core/flow/checkpointer.py:243]: Wondering if we can just fetch the remaining samples based on checkpoint number/save_freq or directly with samples_processed? is it coz of async we are using this logic?

File: src/sdg_hub/core/flow/checkpointer.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        # Filter input dataset to only remaining samples
+        remaining_mask = input_df.apply(tuple, axis=1).isin(remaining_tuples)
+        remaining_indices = input_df[remaining_mask].index.tolist()
+
+        if not remaining_indices:
+            # Return empty dataset with same structure
+            return input_dataset.select([])
+
+        return input_dataset.select(remaining_indices)
+
```
[the remark this answers is not in the document]
Comment (shivchander) [reply src/sdg_hub/core/flow/checkpointer.py:294]: We expect the input dataset to be unique, will note this in comments and update the docs to reflect this

File: src/sdg_hub/core/flow/checkpointer.py (hunk 2 above)
Comment (shivchander) [reply src/sdg_hub/core/flow/checkpointer.py:243]: Yep, async and there  might be samples in between which haven't been processed

**Review by eshwarprasadS — APPROVED** [review]
lgtm thanks for adding this
